pkg/frontend/readpath/queryfrontend/query_series_labels.go

```diff
@@ -9,9 +9,54 @@ import (
 
 	querierv1 "github.com/grafana/pyroscope/api/gen/proto/go/querier/v1"
 	queryv1 "github.com/grafana/pyroscope/api/gen/proto/go/query/v1"
+	"github.com/grafana/pyroscope/pkg/featureflags"
 	"github.com/grafana/pyroscope/pkg/validation"
 )
 
+func (q *QueryFrontend) filterLabelNames(
+	ctx context.Context,
+	c *connect.Request[querierv1.SeriesRequest],
+	labelNames []string,
+	labelSelector string,
+) ([]string, error) {
+	if capabilities, ok := featureflags.GetClientCapabilities(ctx); ok && capabilities.AllowUtf8LabelNames {
+		return labelNames, nil
+	}
+
+	toFilter := make([]string, len(labelNames))
+	copy(toFilter, labelNames)
+
+	// Filter out label names not passing legacy validation if utf8 label names not enabled
+	if len(labelNames) == 0 {
+		// Querying for all label names; must retrieve all label names to then filter out
+		report, err := q.querySingle(ctx, &queryv1.QueryRequest{
+			StartTime:     c.Msg.Start,
+			EndTime:       c.Msg.End,
+			LabelSelector: labelSelector,
+			Query: []*queryv1.Query{{
+				QueryType:  queryv1.QueryType_QUERY_LABEL_NAMES,
+				LabelNames: &queryv1.LabelNamesQuery{},
+			}},
+		})
+		if err != nil {
+			return nil, err
+		}
+		if report != nil {
+			toFilter = report.LabelNames.LabelNames
+		}
+	}
+
+	filtered := make([]string, 0, len(labelNames))
+	for _, name := range toFilter {
+		if _, _, ok := validation.SanitizeLegacyLabelName(name); !ok {
+			level.Debug(q.logger).Log("msg", "filtering out label", "label_name", name)
+			continue
+		}
+		filtered = append(filtered, name)
+	}
+	return filtered, nil
```

Comment:
Is there any reason why we can't call `LabelNames()` and rely on the implementation there?

Reply:
I thought I saw examples of other APIs using this approach for internal calls, but now I can't find that. Happy to use the frontdoor if it works.

Reply:
This approach is filtering out label names that even the client provides. Ok to update to only filter out label names in the case where no label names provided by client.

Also curious: any potential issues with passing a potentially unbounded list of labelNames here?https://github.com/grafana/pyroscope/blob/b2888fae7257f2264df1ca1c739dda132b9b1c9c/pkg/frontend/readpath/queryfrontend/query_series_labels.go#L91-L101